src/d3a/d3a_core/sim_results/file_export_endpoints.py

```diff
@@ -370,50 +371,83 @@ def _accumulate_energy(self, area, child, is_cep: bool):
                     if trade.buyer is child.name:
                         self._total_energy += trade.offer.energy
 
-    def _export_house_pv_self_consumption(self, area):
-        house_pv_device = list()
-        house_load_device = list()
-        trade_by_pv = 0
-        total_energy_bought = 0
-        traded_from_pv = 0
-        if not _is_house_node(area):
-            return
+    def _accumulate_devices(self, area):
         for child in area.children:
-            if isinstance(child.strategy, PVStrategy):
-                house_pv_device.append(child.name)
-            elif _is_load_node(child) or _is_prosumer_node(child):
-                house_load_device.append(child.name)
+            if _is_producer_node(child):
+                self.producer_list.append(child.name)
+            elif _is_load_node(child):
+                self.consumer_list.append(child.name)
+                self.consumer_area_list.append(child.parent)
+                for time, energy in child.strategy.state.desired_energy_Wh.items():
+                    self.total_energy_demanded += energy
+            elif _is_prosumer_node(child):
+                self.ess_list.append(child.name)
+
+            if child.children:
+                self._accumulate_devices(child)
+
+    def _accumulate_self_consumption(self, trade):
+        if trade.seller_origin in self.producer_list and trade.buyer_origin in self.consumer_list:
+            self.total_self_consumption += trade.offer.energy * 1000
+
+    def _accumulate_self_consumption_buffer(self, trade):
+        if trade.seller_origin in self.producer_list and trade.buyer_origin in self.ess_list:
+            self.self_consumption_buffer += trade.offer.energy * 1000
+
+    def _dissipate_self_consumption_buffer(self, trade):
+        if trade.seller_origin in self.ess_list and \
+                (trade.buyer_origin not in self.consumer_list or
+                 trade.buyer_origin not in self.ess_list) and \
+                self.self_consumption_buffer > 0:
+            if (self.self_consumption_buffer - trade.offer.energy * 1000) > 0:
+                self.self_consumption_buffer -= trade.offer.energy * 1000
+            else:
+                self.self_consumption_buffer = 0
+
+    def _accumulate_energy_trace(self):
+        for c_area in self.consumer_area_list:
+            for market in c_area.past_markets:
+                for trade in market.trades:
+                    self._accumulate_self_consumption(trade)
+                    self._accumulate_self_consumption_buffer(trade)
+                    self._dissipate_self_consumption_buffer(trade)
+
+    def _area_self_sufficiency(self, area):
+        self.producer_list = list()
+        self.consumer_list = list()
+        self.consumer_area_list = list()
+        self.ess_list = list()
+        self.total_energy_demanded = 0
+        self.total_self_consumption = 0
+        self.self_consumption_buffer = 0
+
+        self._accumulate_devices(area)
```

Comment:
AFAIU, you dont need to get all the children for every market slot, right? IMO you can derive the lists in the constructor.

Reply:
Please ignore this comment. I just realised that the self sufficiency is not populated for each market but for the whole simulation.